src/boot/efi/vmm.c

```diff
@@ -241,21 +241,15 @@ static const SmbiosHeader *get_smbios_table(uint8_t type, uint64_t *ret_size_lef
                 size -= header->length;
                 p += header->length;
 
-                /* Skip over string table. */
-                for (;;) {
-                        const uint8_t *e = memchr(p, 0, size);
-                        if (!e)
-                                return NULL;
-
-                        if (e == p) {/* Double NUL byte means we've reached the end of the string table. */
-                                p++;
-                                size--;
-                                break;
-                        }
-
-                        size -= e + 1 - p;
-                        p = e + 1;
+                /* Skip chars until a double NUL */
+                while (!(p[0] == '\0' && p[1] == '\0')) {
+                        p += 1;
+                        size -= 1;
                 }
+
+                /* Skip double NUL */
+                p += 2;
+                size -= 2;
```

Comment:
it's not clear to me what kind of SMBIOS table we are supposed to have misparsed so far? can you give an example?

Reply:
@poettering 

The ones that get misparsed are ones that have zero strings, aka where the entire string table area is "\0\0".

In that case the memchr finds the first \0 byte on first iteration of the for loop, then `if (e == p)` is hit, skipping over only the first \0 byte and breaking the loop. This leads to incorrectly assuming the second \0 byte is part of the next smbios table and the whole parsing is off-by-one from then on.

I hit this issue on the SMBIOS table from a ThinkPad X13s.

Reply:
interesting. I thought the string table was supposed to end in a double NUL, no?

Reply:
Yes, it ends in a double NUL, what I am saying is the issue is hit when there are no strings and it is just a double NUL.